federation-js/src/service/printCoreSchema.ts

```diff
@@ -0,0 +1,581 @@
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  if (isScalarType(type)) {
+    return printScalar(type, options);
+  } else if (isObjectType(type)) {
+    return printObject(type, options, context);
+  } else if (isInterfaceType(type)) {
+    return printInterface(type, options);
+  } else if (isUnionType(type)) {
+    return printUnion(type, options);
+  } else if (isEnumType(type)) {
+    return printEnum(type, options);
+  } else if (isInputObjectType(type)) {
+    return printInputObject(type, options);
+  }
+
+  throw Error('Unexpected type: ' + (type as GraphQLNamedType).toString());
+}
+
+function printScalar(type: GraphQLScalarType, options?: Options): string {
+  return printDescription(options, type) + `scalar ${type.name}`;
+}
+
+function printObject(
+  type: GraphQLObjectType,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  const interfaces = type.getInterfaces();
+  const implementedInterfaces = interfaces.length
+    ? ' implements ' + interfaces.map((i) => i.name).join(' & ')
+    : '';
+
+  return (
+    printDescription(options, type) +
+    `type ${type.name}` +
+    implementedInterfaces +
+    // Core addition for printing @join__owner and @join__type usages
+    printTypeJoinDirectives(type, context) +
+    printFields(options, type, context)
+  );
+}
+
+// Core change: print @join__owner and @join__type usages
+function printTypeJoinDirectives(
+  type: GraphQLObjectType,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  const metadata: FederationType = type.extensions?.federation;
+  if (!metadata) return '';
+
+  const { serviceName: ownerService, keys } = metadata;
+  if (!ownerService || !keys) return '';
+
+  // Separate owner @keys from the rest of the @keys so we can print them
+  // adjacent to the @owner directive.
+  const { [ownerService]: ownerKeys = [], ...restKeys } = keys;
+  const ownerEntry: [string, (readonly SelectionNode[])[]] = [
+    ownerService,
+    ownerKeys,
+  ];
+  const restEntries = Object.entries(restKeys);
+
+  return (
+    `\n  @join__owner(graph: ${
+      context?.sanitizedServiceNames[ownerService] ?? ownerService
+    })` +
+    [ownerEntry, ...restEntries]
+      .map(([service, keys = []]) =>
+        keys
+          .map(
+            (selections) =>
+              `\n  @join__type(graph: ${
+                context?.sanitizedServiceNames[service] ?? service
+              }, key: ${printStringLiteral(printFieldSet(selections))})`,
+          )
+          .join(''),
+      )
+      .join('')
+  );
+}
+
+function printInterface(
+  type: GraphQLInterfaceType,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  return (
+    printDescription(options, type) +
+    `interface ${type.name}` +
```

Comment:
Is there potentially a `printTypeJoinDirectives` missing here (see `printObject`)? I think the state of interfaces as entities is a bit rocky at the moment but still something we want to support, and probably should at the printer level.

Reply:
Yeah, I didn't realize we kept track of this already, but printing `join__type` on interfaces would make avoiding type explosion less error prone / more effective, so that's definitely worth it.